[ONNX][TypePromo] Explicit type promotion pass

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104491
* #104720
* #104229
* __->__ #104064

This PR adds the `ExplicitTypePromotionPass` that does an fx graph to fx graph transformation
explicitly adding cast nodes into the graph to emulate the PyTorch type promotion behavior.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR

BowenBao committed:
[ONNX][TypePromo] Explicit type promotion pass

[ghstack-poisoned]

BowenBao committed:
rebase on "[ONNX][TypePromo] Explicit type promotion pass"




[ghstack-poisoned]

BowenBao committed:
Update rule set; Support node re-dispatch. Support tuple/list fx arg on "[ONNX][TypePromo] Explicit type promotion pass"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][TypePromo] Explicit type promotion pass"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][TypePromo] Explicit type promotion pass"




[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -556,7 +490,7 @@
         reason="FIXME: After https://github.com/microsoft/onnxruntime/issues/15446 is fixed",
     ),
-    xfail(
+    skip(
```
After type promotion this test starts to fail a few subtests with unexpected pass, which appears to be incosistent with the taggex xfail reason.

BowenBao replied:
Actually more likely due to the bumping onnxscript version in this PR, so unrelated to type promotion. The inconsistency concern still holds.

BowenBao committed:
Update rule docs and comments; Relax dispatcher check when producing multiple ops on "[ONNX][TypePromo] Explicit type promotion pass"


This PR adds the `ExplicitTypePromotionPass` that does an fx graph to fx graph transformation
explicitly adding cast nodes into the graph to emulate the PyTorch type promotion behavior.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR


[ghstack-poisoned]

BowenBao commented on .ci/docker/common/install_onnx.sh:
```diff
@@ -24,7 +24,7 @@ pip_install \
 
 # TODO: change this when onnx-script is on testPypi
-pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@7e131c578f290ffad1f26bacda11a83daf5476ba"
+pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@2bb3e9f2d094912f81cb63cecb412efb14c65738"
```
note: for missing `aten.tensor` op support that is used when promoting python number to tensor.

titaiwangms commented on torch/onnx/_internal/diagnostics/_rules.py:
```diff
@@ -844,6 +864,31 @@ class _POERules(infra.RuleCollection):
+    fx_node_explicit_type_promotion: _FxNodeExplicitTypePromotion = dataclasses.field(
+        default=_FxNodeExplicitTypePromotion.from_sarif(
+            **{
+                "id": "FXE0013",
```
CONFLICT!!! FXE0013 has been taken. By the way, what does FXE mean? If it only means FX exporter, do we want to utilize it to further separate rules under FX exporter?

thiagocrepaldi replied:
good point, do we have any way to find this conflict? maybe a test that runs on ci?

BowenBao replied:
don't worry, it is merge conflicts 🤣

titaiwangms commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -20,15 +21,26 @@
 logger = logging.getLogger(__name__)
 
+# TODO(bowbao): move to type utils.
+_SCALAR_TYPE_TENSOR_DTYPE_MAP: Mapping[type, torch.dtype] = {
```
Also please make it accessible through a stateless function or class.

titaiwangms replied:
Maybe we can unify the need of torch.dtype conversion into something like JitScalarType system now. The needs are increasing.

justinchuby replied:
> Also please make it accessible through a stateless function or class.

If it is private I think it's fine for now

BowenBao replied:
Should we make function accessor `get(..., None)` style? I'm not sure but it is used here so I'm leaving out type_utils in this PR.

titaiwangms commented on torch/onnx/_internal/diagnostics/rules.yaml:
```diff
@@ -313,6 +313,32 @@
+      the PyTorch behavior. This diagnostic tracks the process of node level explicit type promotion.
+
+      A set of type promotion rules supported by this pass is materialized under `torch/onnx/_internal/fx/passes/type_promotion.py`.
+      Checkout [Note: Add type promotion rule] how to update or add new type promotion rules.
```
If it's not rephrased by chatGPT yet, you can try with it. Not necessarily, but it does give me a decent docs.

titaiwangms commented on torch/onnx/_internal/diagnostics/rules.yaml:
```diff
@@ -313,6 +313,32 @@
+      Checkout [Note: Add type promotion rule] how to update or add new type promotion rules.
+  message_strings:
+    default:
+      text: "Performing explicit type promotion for node {target}. "
```
I forgot to ask, is {target} decided by the decorated function?

BowenBao commented:
I will update new diagnostic rule id prior to merging.

titaiwangms commented on torch/onnx/_internal/diagnostics/rules.yaml:
```diff
@@ -313,6 +313,32 @@
+  full_description:
+    text: "Determine if type promotion is required for the FX node. Explicitly insert cast nodes if needed."
+    markdown: |
+      This diagnostic tracks the process of node level explicit type promotion. Implicit [type promotion](https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc) is a
```
I think we should have a better explanation here on "it's needed to be explicit in ONNX, because it's implicit in PyTorch.". It seems to me the explanation here is a bit confusing.

titaiwangms replied:
The same idea goes to the name itself. When we say: fx-node-explicit-type-promotion, it means we need explicit type promotion for ONNX. But it could be just me...

BowenBao replied:
Is `fx-node-insert-explicit-type-promotion` better? Or just `fx-node-insert-type-promotion`

BowenBao replied:
I'm leaving `explicit` in most of the places and all the names.

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -635,6 +635,10 @@ def pre_export_passes(
 
+    # ONNX does not support concept of (implicit) type promotion.
+    # Insert type casts explicitly where needed.
+    module = passes.ExplicitTypePromotionPass(diagnostic_context, module).run()
```
Is there a convention for naming the passes? Looks like the surrounding ones start with verbs

BowenBao replied:
good catch, I will name it as `InsertExplicitTypePromotion`.

BowenBao replied:
Or would you prefer `InsertTypePromotion`?

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -55,6 +67,16 @@ class TypePromotionSnapshot:
 
 
+@_beartype.beartype
+def _get_fake_tensor_val(node: torch.fx.Node) -> fake_tensor.FakeTensor:
```
Add a brief on how it should be used?

BowenBao replied:
syntactic sugar, maybe call it `_fake_tensor_from_meta`?

justinchuby replied:
or `_fake_tensor_from_node_meta`

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -920,3 +947,378 @@ def get_rule(
+        type_promotion_table: Type promotion table.
+
+    Returns:
+        Type promotion rule for the node.
```
Document None as well (when it happens)?

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -920,3 +947,378 @@ def get_rule(
+    return rule
+
+
+class _OpTraceDispatchMode(_python_dispatch.TorchDispatchMode):
```
Just for my learning. What does it do/why do we need it? Thanks!

BowenBao replied:
added docstring

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -920,3 +947,378 @@ def get_rule(
+
+
+@_beartype.beartype
+def find_compatible_op_overload(
```
So this is like a tiny dispatcher?

thiagocrepaldi replied:
Indeed, confusing. Any chance we could reuse or extend the current our dispatcher for this?

BowenBao replied:
I'll expand the docstring. This is indeed a tricky problem. I have tried to solve in multiple different ways. 
The current one is the most elegant and simple solution.

titaiwangms replied:
Looks like when we type promote(change) args/kwargs, the opoverloadpacket needs to be dispatched again in FX graph.

I wonder if they have a schema matching system that we could use.

BowenBao replied:
> I wonder if they have a schema matching system that we could use.

afaik there's only a lower level api that checks if two schemas are compatible to each other, and it is painful to use in this scenario since we need to create a new schema from args/kwargs and check existing schemas for matches. And furthermore due to the ordering of check the schema we found might disagree with torch dispatcher. I found it easier to let `__torch_dispatch__` tell us which opoverload it dispatched to in the end.

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -920,3 +947,378 @@ def get_rule(
+
+    Raises:
+        RuntimeError: If no compatible op overload is found.
+        AssertionError: If compatible op overload is not an instance of torch._ops.OpOverload.
+        AssertionError: If op is being decomposed.
```
Do we expect them to happen with valid user input? If not I don't think we need to document them here. Instead I would document as invariants under the summary, before "Returns".

justinchuby replied:
If yes we should raise something that's not assertion error

BowenBao replied:
I'll drop them from docstring.

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -920,3 +947,378 @@ def get_rule(
+    return new_op_overload
+
+
+@dataclasses.dataclass
```
I am here. Taking a break for now.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -55,6 +67,16 @@ class TypePromotionSnapshot:
 
+@_beartype.beartype
+def _get_fake_tensor_val(node: torch.fx.Node) -> fake_tensor.FakeTensor:
+    val = node.meta.get("val", None)
```
move inside the if? it makes no sense to set it to none if it will railse on next if

BowenBao replied:
Not sure how to do it without accessing node.meta multiple times.
With walrus I can make 1 line shorter but I don't feel it's necessary.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -68,8 +90,11 @@ def __init__(
     ):
         self.namespace = namespace
         self.op_name = op_name
+        # Positions of args to promote.
```
maybe create official docs strings with `Args:` section?

BowenBao replied:
done

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -920,3 +947,378 @@ def get_rule(
+
+
+@dataclasses.dataclass
+class NodeMeta:
```
Are we keeping another "meta" structure or using this class to update the fx node meta field? please elaborate on the docstring

BowenBao replied:
It is an attempt, but I'll remove it because this is not actually used.

It is painful to maintain this node.meta dict. The idea is to create an ONNX meta abstraction to be used with our fx.Node creation api such that when people create new nodes they don't forget to set certain meta info: every field is required.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -920,3 +947,378 @@ def get_rule(
+            setattr(node.meta, field.name, getattr(self, field.name))
+
+
+class ExplicitTypePromotionPass(_pass.Transform, torch.fx.Interpreter):
```
wow, multiple inheritance. We certainly need a lot of docstring here to describe what and how things are being done. examples appreciated :)

BowenBao replied:
I will separate it. Thinking again, it is better at least in this PR to keep them separated.

thiagocrepaldi reviewed (commented):
This PR has a lot of complexity and very little documentation. Hard to follow

BowenBao committed:
Add comments & cleanup refactor on "[ONNX][TypePromo] Explicit type promotion pass"


This PR adds the `ExplicitTypePromotionPass` that does an fx graph to fx graph transformation
explicitly adding cast nodes into the graph to emulate the PyTorch type promotion behavior.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR


[ghstack-poisoned]

BowenBao commented on torch/csrc/onnx/diagnostics/generated/rules.h:
```diff
@@ -19,14 +19,12 @@ enum class Rule : uint32_t {
   /**
-   * @brief Missing symbolic function for custom PyTorch operator, cannot
-   * translate node to ONNX.
+   * @brief Missing symbolic function for custom PyTorch operator, cannot translate node to ONNX.
```
this is strange 🤔 wonder if lint config changed

BowenBao replied:
forgot to lintrunner init

BowenBao committed:
Update on "[ONNX][TypePromo] Explicit type promotion pass"


This PR adds the `ExplicitTypePromotionPass` that does an fx graph to fx graph transformation
explicitly adding cast nodes into the graph to emulate the PyTorch type promotion behavior.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR


[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -933,3 +968,438 @@ def get_rule(
+        fx_arg: torch.fx.node.Argument,
+        dtype: Optional[torch.dtype],
+    ) -> torch.fx.node.Argument:
+        """Promote fx_arg to dtype if necessary."""
```
Does the method check for necessity or do we call it if necessary? If the former, consider `_maybe_promote_arg`?

BowenBao replied:
I can rename both with `_maybe` prefix.

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -933,3 +968,438 @@ def get_rule(
+        node: torch.fx.Node,
+        rule: TypePromotionRule,
+    ) -> torch.fx.Node:
+        """Promote node inputs and outputs according to type promotion rule."""
```
Just for my understanding, only _promote_node will use _promote_arg right?

BowenBao replied:
Correct.

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -933,3 +968,438 @@ def get_rule(
+        rule=diagnostics.rules.fx_node_insert_type_promotion,
+        level=diagnostics.levels.NONE,
+    )
+    def run_node(self, node: torch.fx.Node) -> Any:
```
It'd be helpful to explain the high-level calling logic in English here I think: Is it `run_node -> _promote_node -> _promote_args -> _create_node -> _run_node_and_update_meta_val -> rerun_node`? When we do rerun, does it happen exactly once for each node?

BowenBao replied:
Yea except for swapping this order here `rerun_node -> _run_node_and_update_meta_val`. Each node is guaranteed to run at least once. We can change the code to run every node exactly once, but it adds more complexity and I don't feel it is necessary since this is always running in fake mode.

I'll update the docstring with high level logic of things to do.

justinchuby replied:
> Each node is guaranteed to run at least once.

Sounds good! I would just make this clear in the docs

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -933,3 +968,438 @@ def get_rule(
+        type_promotion_table: Optional[TypePromotionTable] = None,
+    ):
+        super().__init__(diagnostic_context, module)
+        self.type_promotion_table = type_promotion_table or TypePromotionTable()
```
this can be inferred from `self.interpreter.type_promotion_table`.

BowenBao commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -933,3 +968,438 @@ def get_rule(
+        By default `Interpreter.run_node` does not update `node.meta["val"]`.
+        """
+        out = super().run_node(node)
+        self.env[node] = out
```
ok "Update interpreter env state with new output value."

BowenBao commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -933,3 +968,438 @@ def get_rule(
+        ), f"Expected OpOverload, got {type(target)}"
+        node.target = find_compatible_op_overload(target.overloadpacket, args, kwargs)
+
+        new_node_val = self._run_node_and_update_meta_val(node)
```
No because this is not a new node. Frankly this is case by case. In this pass for modified nodes we only need to update 'val' in 'meta'. But if you prefer I can simplify the api to always copy all other meta info.

Largely this is the problem with freestyle design of `fx.Node` meta. Devs need to be super aware of its content to correctly maintain it. And meta is very important as it serves as source info for numerous high level apis.

BowenBao replied:
I will isolate the change in a follow up PR.

BowenBao replied:
https://github.com/pytorch/pytorch/pull/104720

BowenBao committed:
Update on "[ONNX][TypePromo] Explicit type promotion pass"


This PR adds the `ExplicitTypePromotionPass` that does an fx graph to fx graph transformation
explicitly adding cast nodes into the graph to emulate the PyTorch type promotion behavior.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR


[ghstack-poisoned]

BowenBao committed:
rebase, CI failing left-right-center unrelated on "[ONNX][TypePromo] Explicit type promotion pass"


This PR adds the `ExplicitTypePromotionPass` that does an fx graph to fx graph transformation
explicitly adding cast nodes into the graph to emulate the PyTorch type promotion behavior.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR


[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -f "unrelated cudnn failure"
